sfdoc/publish/salesforce.py

```diff
@@ -173,53 +245,90 @@ def get_preview_url(self, ka_id, online=False):
         return preview_url
 
-    def process_article(self, html, bundle):
+    def process_draft(self, html, bundle):
         """Create a draft KnowledgeArticleVersion."""
+        logger = get_logger(bundle)
 
         # update links to draft versions
-        html.update_links_draft(self.get_base_url())
+        html.update_links_draft(bundle.docset_id, self.get_base_url())
 
         # query for existing article
-        result_draft = self.query_articles(html.url_name, 'draft')
-        result_online = self.query_articles(html.url_name, 'online')
+        result_draft = self.find_articles_by_name(html.url_name, 'draft')
+        result_online = self.find_articles_by_name(html.url_name, 'online')
 
-        if result_draft['totalSize'] == 1:
+        if len(result_draft) == 1:
             # draft exists, update fields
-            kav_id = result_draft['records'][0]['Id']
+            kav_id = result_draft[0]['Id']
             self.update_draft(kav_id, html)
-            if result_online['totalSize'] == 1:
+            if len(result_online) == 1:
                 # published version exists
                 status = Article.STATUS_CHANGED
             else:
                 # not published
                 status = Article.STATUS_NEW
-        elif result_online['totalSize'] == 0:
+            logger.info("Draft updated with DB status %s, %s", status, html.url_name)
+        elif len(result_online) == 0:
             # new draft, new article
             kav_id = self.create_article(html)
             status = Article.STATUS_NEW
-        elif result_online['totalSize'] == 1:
+            logger.info("Draft created with DB status %s, %s", status, html.url_name)
+        elif len(result_online) == 1:
             # new draft of existing article
-            record = result_online['records'][0]
+            record = result_online[0]
             # check for changes in article fields
             if html.same_as_record(record):
                 # no update
+                logger.info("Draft did not change: skipping: %s", html.url_name)
                 return
             # create draft copy of published article
             kav_id = self.create_draft(record['KnowledgeArticleId'])
             self.update_draft(kav_id, html)
             status = Article.STATUS_CHANGED
+            logger.info("New draft of published article, status %s, %s", status, html.url_name)
 
         self.save_article(kav_id, html, bundle, status)
 
+    #  As with all caches, be careful with this one.
+    #  Several things have bitten me with it already.
+    #  1. You MUST specify a publish_status in the query. Draft records go missing if you don't.
+    #  2. If you don't manage it as a class variable, you could get two different out-of-sync views of the data.
+    #  3. If the data changes remotely while you have an open Salesforce docset view, you're hooped, obviously.
+    def article_info_cache(self, publish_status, **filters):
```

Comment:
Can we name this something with a verb like `query_articles_cached`? Would make its function clearer when reading the code that calls it.

Reply:
40142c5

Reply:
Now the data is back to being a noun.

81bf7c7